### PR TITLE
Troubleshooting machine connections

### PR DESCRIPTION
**Pull request by petervo**
Still a WIP, more authentication error handling and support for ports and users still to come.

### CODE REVIEW

File: src/ws/cockpitwebservice.c [hunk 1]
```diff
@@ -1103,12 +1131,25 @@ on_session_closed (CockpitTransport *transport,
...
-                                           "problem", problem,
-                                           "host-key", key,
-                                           "host-fingerprint", fp,
-                                           NULL);
+                  object = build_json ("command", "close",
+                                       "channel", channel,
+                                       "problem", public_problem,
+                                       "host-key", key,
+                                       "host-fingerprint", fp,
+                                       "host-key-failure", hostkey_failure,
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: Where is this documented? Also what is the reasoning behind not just using varying problem codes?

File: src/ws/cockpitsshtransport.c [hunk 2]
```diff
@@ -83,29 +83,54 @@ typedef struct {
...
-    g_string_append (string, "password ");
-  if (methods & SSH_AUTH_METHOD_PUBLICKEY)
-    g_string_append (string, "public-key ");
-  if (methods & SSH_AUTH_METHOD_HOSTBASED)
-    g_string_append (string, "host-based ");
-  if (methods & SSH_AUTH_METHOD_GSSAPI_MIC)
-    g_string_append (string, "gssapi-mic ");
+    g_string_append (string, auth_method_description(methods));
+
+  for (i = 0; i < 4; i++)
```
Comment (stefwalter) [inline src/ws/cockpitsshtransport.c]: How about using ```G_N_ELEMENTS(check)```

File: src/ws/cockpitsshtransport.c [hunk 3]
```diff
@@ -83,29 +83,54 @@ typedef struct {
...
-    g_string_append (string, "none ");
-  if (methods & SSH_AUTH_METHOD_PASSWORD)
-    g_string_append (string, "password ");
-  if (methods & SSH_AUTH_METHOD_PUBLICKEY)
-    g_string_append (string, "public-key ");
-  if (methods & SSH_AUTH_METHOD_HOSTBASED)
-    g_string_append (string, "host-based ");
-  if (methods & SSH_AUTH_METHOD_GSSAPI_MIC)
-    g_string_append (string, "gssapi-mic ");
+    g_string_append (string, auth_method_description(methods));
```
Comment (stefwalter) [inline src/ws/cockpitsshtransport.c]: Needs a space before paren.

File: src/ws/cockpitsshtransport.h [hunk 4]
```diff
@@ -33,15 +33,17 @@ G_BEGIN_DECLS
 typedef struct _CockpitSshTransport        CockpitSshTransport;
 typedef struct _CockpitSshTransportClass   CockpitSshTransportClass;
 
-GType               cockpit_ssh_transport_get_type              (void) G_GNUC_CONST;
+GType               cockpit_ssh_transport_get_type                (void) G_GNUC_CONST;
```
Comment (stefwalter) [inline src/ws/cockpitsshtransport.h]: Typically we don't reindent the headers, just add another section down below that is further indented.

File: tools/mustache-bundle [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    input = ops.args.shift();
+
+    fs.readFile(input, { encoding: "utf-8" }, function(err, data) {
+        if (err)
+            fatal(err.message);
+
+        var name = path.basename(input, path.extname(input));
+        templates[name] = data;
+	    step();
```
Comment (stefwalter) [inline tools/mustache-bundle]: Indentation

File: src/ws/cockpitwebservice.c [hunk 6]
```diff
@@ -1119,43 +1160,120 @@ on_session_closed (CockpitTransport *transport,
...
+  g_free(user_arg);
+}
+
 static CockpitSession *
 lookup_or_open_session_for_host (CockpitWebService *self,
                                  const gchar *host,
                                  const gchar *host_key,
-                                 CockpitCreds *creds,
+                                 const gchar *user,
+                                 const gchar *password,
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: What's the logic behind this change?

File: src/ws/cockpitwebservice.c [hunk 7]
```diff
@@ -1119,43 +1160,120 @@ on_session_closed (CockpitTransport *transport,
...
+    }
+  else
+    {
+      *hostname = g_strdup (host_arg);
+    }
+
+  *username = g_strdup (user_arg);
+  *port = port_num;
+
+  g_free(host_arg);
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: Needs space here.

File: src/ws/cockpitwebservice.c [hunk 8]
```diff
@@ -1119,43 +1160,120 @@ on_session_closed (CockpitTransport *transport,
...
+
+  tmp = strrchr (host, ':');
+  if (tmp)
+    {
+      tmp_num = g_ascii_strtoull (tmp + 1, &end, 10);
+      if (end[0] == '\0' && tmp_num < G_MAXUSHORT)
+        {
+          port_num = (guint) tmp_num;
+          host_length = host_length - strlen (tmp);
+        }
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: Should we print a g_message() when we encounter an invalid port?

File: src/ws/cockpitwebservice.c [hunk 9]
```diff
@@ -1119,43 +1160,120 @@ on_session_closed (CockpitTransport *transport,
...
+  gsize host_offset = 0;
+  gsize host_length = strlen (host);
+
+  tmp = strrchr (host, '@');
+  if (tmp && tmp[0] != host[0])
+    {
+      user_arg = g_strndup (host, tmp - host);
+      host_offset = strlen (user_arg) + 1;
+      host_length = host_length - host_offset;
+    }
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: Should we print a g_message when we encounter an invalid user name?

File: pkg/shell/templates.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    "color-picker",
+    "host-key",
+    "not-supported",
+    "sync-users"
+];
+
+var deps = templates.map(function(template) {
+    return "data!shell/templates/" + template + ".html";
+});
+deps.unshift("base1/mustache");
```
Comment (stefwalter) [inline pkg/shell/templates.js]: This seems like an overly complicated way of avoiding typing a few extra characters. I mean the strings are already hard coded above.

File: pkg/shell/templates.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+define(deps, function(mustache) {
+    "use strict";
+    var module = {};
+
+    for (var i = 0; i < templates.length; i++) {
+        var data = arguments[i+1];
+        module[templates[i]] = data;
+        if (data)
+            mustache.parse(data);
+    }
```
Comment (stefwalter) [inline pkg/shell/templates.js]: I'm not convinced this is a good way to approach this. Essentially we're reinventing AMD modules at another level. Why not just load the individual templates as modules. We could make a mustache loader in require-loaders.js.

File: pkg/shell/machines.js [hunk 12]
```diff
@@ -100,12 +100,19 @@ define([
...
+                machine = sync(old_machine, content[host], overlay[host]);
 
                 /* Fill in defaults */
                 machine.key = host;
                 if (!machine.address)
                     machine.address = host;
+
+                machine.connection_string = self.generate_connection_string(machine.user,
+ machine.port,
+ machine.address);
```
Comment (stefwalter) [inline pkg/shell/machines.js]: Indentation.

File: pkg/shell/machines.js [hunk 13]
```diff
@@ -205,16 +196,26 @@ define([
...
-                    .fail(function(ex) {
-                        console.warn("couldn't set pretty host name: " + ex);
-                    });
+
+                var conn_to = host;
+                if (machine)
+                    conn_to = machine.connection_string;
+
+                if (!machine || machine.label !== values.label) {
+
```
Comment (stefwalter) [inline pkg/shell/machines.js]: Unnecessary line?

File: pkg/shell/machine-dialogs.js [hunk 14]
```diff
@@ -0,0 +1,890 @@
...
+    "shell/machines",
+    "shell/templates",
+    "shell/credentials",
+    'translated!base1/po',
+    "base1/patterns",
+    "base1/bootstrap-select",
+], function($, cockpit, mustache, machines, templates, credentials, po) {
+    "use strict";
+
+    cockpit.locale(po);
```
Comment (stefwalter) [inline pkg/shell/machine-dialogs.js]: This only needs to happen once per HTML page. Is this the right place?

**Comment by stefwalter** [comment]
Should we say 'Add Machine to Dashboard' to make it clear what's going on?

 * The color box should be smaller.
 * The colors in the dropdown should not wrap.

screenshot from 2015-11-17 10-21-34

**Comment by stefwalter** [comment]
The alignment in this dialog is off. The content is closer to the edges than the buttons.

screenshot from 2015-11-17 10-24-11

**Comment by stefwalter** [comment]
The dialog says 'Add Machine: Incorrect Host Key' when I have never interacted with this host using Cockpit before. The entire contents of my known_hosts are:

```
null
127.0.0.1 ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIKr9O3W7oJEPhz0ty8uopxBNkA2siNl8vTLY761jJ+SO
```

The dialog should be renamed to 'Unknown Host Key' or something like that.

**Comment by stefwalter** [comment]
Some fixes to this dialog:

 * There's a dark border at the top of this box. We don't need that.
 * There should be spaces in between the group names.
 * Lets make the password use bullets instead of stars.

screenshot from 2015-11-17 10-26-55

**Comment by stefwalter** [comment]
We shouldn't have a generic 'Troubleshoot' button:

 * The name of the button should be taken from the error condition. In this case 'Add Machine'

screenshot from 2015-11-17 10-29-02

File: pkg/shell/templates.js (hunk 11 above)
Comment (petervo) [inline pkg/shell/templates.js]: The idea is that to modules that use templates.js things look exactly the same in both debug or complied mode. It's an optimization so that in debug mode the files can be split out but in  compiled mode all the templates are loaded with one call. Do you want to get rid of that and make all the templates one html file or load multiple files separately?

**Comment by petervo** [comment]
On 11/17/2015 01:29 AM, Stef Walter wrote:
> We shouldn't have a generic 'Troubleshoot' button:

What about the errors that don't have a recovery action ie no-cockpit or 
some of the host key ones.

**Comment by stefwalter** [comment]
> What about the errors that don't have a recovery action ie no-cockpit or
some of the host key ones.

Should we just display the failure inline on the blank state screen in that case?

**Comment by stefwalter** [comment]
If having a Troubleshoot button makes things much easier. Then i'm fine with leaving it that way for this pull request. We'll need to make changes to this later anyway

**Comment by petervo** [comment]
> Should we say 'Add Machine to Dashboard' to make it clear what's going on?

Do you want to change this in both contexts or only when clicked from 
the dashboard?

Also right now i'm using a prefix when adding a machine when you get 
branched to other dialogs

Add Machine: Dialog title

Do you want to change that prefix too or is that too long?

**Comment by stefwalter** [comment]
> Also right now i'm using a prefix when adding a machine when you get
branched to other dialogs

Does the prefix really matter? Even while adding a machine? I would say no.

**Comment by stefwalter** [comment]
> The idea is that to modules that use templates.js things look exactly the same in both debug or complied mode. It's an optimization so that in debug mode the files can be split out but in compiled mode all the templates are loaded with one call. Do you want to get rid of that and make all the templates one html file or load multiple files separately?

That's exactly what we already do in the other modules. For example see the ```pkg/realmd``` component. requirejs already provides exactly this ability, so it's not clear why we're putting another layer on top of it.

**Comment by petervo** [comment]
Updated

**Comment by stefwalter** [comment]
Size of color area in dialog should be 26px high. It should have no margin above it (no 10px margin).

It seems like when I add a server on the dashboard, and then disconnect it (suspend that laptop) I never see the server go "red" or into failure mode. Is that intentional? When I click on such a machine i get a blank screen. No curtains or anything.

**Comment by stefwalter** [comment]
screenshot from 2015-11-19 21-34-23

This needs:

 * More space between question and field (10 15 px?)
 * The port field should be 100px wide.
 * The dialog should be modal-sm

File: pkg/base1/cockpit.js [hunk 15]
```diff
@@ -2366,6 +2366,8 @@ function full_scope(cockpit, $, po) {
             return _("The server refused to authenticate using any supported methods.");
         else if (problem == "unknown-hostkey")
             return _("Untrusted host");
+        else if (problem == "changed-hostkey")
+            return _("Host key changed");
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Shouldn't this be 'invalid-hostkey' and "Host key is incorrect" or "Host key is invalid". "Changed" doesn't sound like a problem to the user.

File: pkg/shell/Makefile.am [hunk 16]
```diff
@@ -6,6 +6,19 @@ nodist_shell_DATA = \
...
+    pkg/shell/templates/add-machine.html \
+    pkg/shell/templates/auth-failed.html \
+    pkg/shell/templates/change-auth.html \
+    pkg/shell/templates/change-port.html \
+    pkg/shell/templates/changed-hostkey.html \
+    pkg/shell/templates/color-picker.html \
+    pkg/shell/templates/not-supported.html \
+    pkg/shell/templates/sync-users.html \
+    pkg/shell/templates/unknown-hostkey.html \
+	$(NULL)
```
Comment (stefwalter) [inline pkg/shell/Makefile.am]: These should be indented with a tab.

File: pkg/shell/machine-dialogs.js [hunk 17]
```diff
@@ -0,0 +1,922 @@
...
+            var button = dialog.get_sel(".btn-primary");
+            if (addr === "") {
+                disabled = true;
+            } else if (!machines.allow_connection_string &&
+                       (addr.indexOf('@') > -1 || addr.indexOf(':') > -1)) {
+                ex = new Error(_("This version of cockpit-ws does not support connecting to a host with an alternate user or port"));
+            } else if (addr.search(/\s+/) === -1) {
+                disabled = false;
+            } else {
+                ex = new Error(_("IP address or host name cannot contain whitespace."));
```
Comment (stefwalter) [inline pkg/shell/machine-dialogs.js]: "The IP address or hostname

File: pkg/shell/machine-dialogs.js [hunk 18]
```diff
@@ -0,0 +1,922 @@
...
+                if (ex.problem == "no-host")
+                    ex = cockpit.message(ex);
+                dialog.render_error(ex);
+            });
+        }
+
+        self.load = function() {
+            var limit_el = $(".machine-limit");
+            if (!limit_el.length)
+                limit_el = $('.machine-limit', window.parent.document);
```
Comment (stefwalter) [inline pkg/shell/machine-dialogs.js]: Access to parent can fail with an exception.

File: pkg/shell/templates/auth-failed.html [hunk 19]
```diff
@@ -0,0 +1,31 @@
+<div class="modal-header">
+    <h4 class="modal-title" translatable="yes">Authentication Failed</h4>
+</div>
+<div class="modal-body">
+    <div translatable="yes">
+    {{#supported}}
+    <div translatable="yes">
```
Comment (stefwalter) [inline pkg/shell/templates/auth-failed.html]: Unfortunately we cannot translate things with tags in them. I think the tags get killed.

File: pkg/shell/templates/change-port.html [hunk 20]
```diff
@@ -0,0 +1,47 @@
+<div class="modal-header">
+    <button type="button" class="close" data-dismiss="modal"
+        aria-hidden="true">&times;</button>
+    <h4 class="modal-title" translatable="yes">Could not contact {{host}}</h4>
+</div>
+
+<div class="modal-body">
+    <div translatable="yes">
```
Comment (stefwalter) [inline pkg/shell/templates/change-port.html]: I don't think we can use this with a translatable attribute mixed with mustache elements.

File: pkg/shell/templates/unknown-hostkey.html [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+<div class="modal-body">
+    {{#key}}
+        <p translatable="yes">
+            The authenticity of host <strong>{{host}}</strong> can't be established. Are you sure you want to continue connecting?
+        </p>
+        <div>
+        <table class="cockpit-form-table">
+            <tr>
+                <td translatable="yes">Fingerprint</td>
+                <td><pre style="margin: 0">{{key}} </pre></td>
```
Comment (stefwalter) [inline pkg/shell/templates/unknown-hostkey.html]: Can we include the style in a CSS class?

File: src/ws/cockpitsshtransport.c [hunk 22]
```diff
@@ -333,170 +359,273 @@ verify_knownhost (CockpitSshData *data)
...
-  int methods;
+  const gchar *password = NULL;
   int rc;
 
-  problem = "authentication-failed";
-
-  rc = ssh_userauth_none (data->session, NULL);
-  if (rc == SSH_AUTH_ERROR)
+  password = cockpit_creds_get_password (data->creds);
+  rc = ssh_userauth_password (data->session, NULL, password);
```
Comment (stefwalter) [inline src/ws/cockpitsshtransport.c]: Does this call work with a null password?

Comment (stefwalter) [inline src/ws/cockpitsshtransport.c]: If password is supposed to be non-null (based on a condition in a calling function in the same file) then we should assert it.

File: src/ws/cockpitwebservice.c [hunk 23]
```diff
@@ -1669,6 +1787,7 @@ cockpit_web_service_ensure_transport (CockpitWebService *self,
   return session->transport;
 }
 
+
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: Unnecessary change.

**Comment by petervo** [comment]
>   * The dialog should be modal-sm

You sure about this? I am concerned it would be weird if the dialog 
changes sizes as you are working through problems. Ei after the fixing 
the port authentication fails.

**Comment by stefwalter** [comment]
> You sure about this? I am concerned it would be weird if the dialog
changes sizes as you are working through problems. Ei after the fixing
the port authentication fails.

Very good point. Lets leave it the way it is.
